lib/SILOptimizer/SILCombiner/SILCombinerApplyVisitors.cpp

```diff
@@ -573,20 +573,31 @@ SILCombiner::optimizeConcatenationOfStringLiterals(ApplyInst *AI) {
 /// If the value is copied from another stack location, \p isCopied is set to
 /// true.
 static SILValue getAddressOfStackInit(AllocStackInst *ASI,
-                                      SILInstruction *ASIUser,
-                                      bool &isCopied) {
+                                      SILInstruction *ASIUser, bool &isCopied,
+                                      DominanceInfo *DT,
+                                      SILInstruction *OrigASIUser = nullptr) {
   SILInstruction *SingleWrite = nullptr;
   // Check that this alloc_stack is initialized only once.
   for (auto Use : ASI->getUses()) {
     auto *User = Use->getUser();
 
     // Ignore instructions which don't write to the stack location.
     // Also ignore ASIUser (only kicks in if ASIUser is the original apply).
-    if (isa<DeallocStackInst>(User) || isa<DebugValueAddrInst>(User) ||
-        isa<DestroyAddrInst>(User) || isa<WitnessMethodInst>(User) ||
-        isa<DeinitExistentialAddrInst>(User) ||
-        isa<OpenExistentialAddrInst>(User) ||
-        User == ASIUser) {
+    if (isa<DebugValueAddrInst>(User) || isa<WitnessMethodInst>(User) ||
+        isa<OpenExistentialAddrInst>(User) || User == ASIUser) {
+      continue;
+    }
+    // Ignore instructions which destroy the stack lcoation.
+    // *as long as* they do not appear before the User!
+    // Else we will have use-after-free
+    if (isa<DeallocStackInst>(User) || isa<DestroyAddrInst>(User) ||
+        isa<DeinitExistentialAddrInst>(User)) {
+      if (OrigASIUser && DT->properlyDominates(User, OrigASIUser)) {
+        return nullptr;
+      }
+      if (DT->properlyDominates(User, ASIUser)) {
+        return nullptr;
+      }
```

Comment:
I agree with Roman's comment about the dominator tree. Maybe - as a quick fix - you can just do a simple check without the dominator tree: If the destroy is in the same block as the OrigASIUser, then check if the destroy is before or after it. If it's in a different block, just bail. I assume that in most cases everything is in the same block.

Reply:
Done